### ISSUE TITLE
[unstable option] comment_width

### ISSUE THREAD

**Issue by scampi** [rust-lang/rustfmt#3349]
Tracking issue for unstable option: comment_width

**Comment by carols10cents** [comment]
Not sure if this is the right place to add this comment or not; feel free to tell me to go somewhere else. 

I was surprised to find out that the default value for `comment_width` is 80, while the default value for `max_width` is 100. I would have expected the defaults to be the same.

So before `comment_width` is stabilized, I'd like to advocate for reconsidering the default value.

Do text editors display/enforce different widths for comments vs code? I use TextMate, and it just has one wrap column setting.

Examples:

<details>

What looks "good" to me:

```rust
// This line of code is just about width 100
fn evaluate_something_test(something: &Something, number: usize) -> Result<BTreeMap, CustomError> {
    // This comment is wrapped to 100. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed 
    // do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis 
    // nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute...
    unimplemented!();
    
    // This is what I would expect. irure dolor in reprehenderit in voluptate velit esse cillum 
    // dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in 
    // culpa qui officia deserunt mollit anim id est laborum.
}
```

What looks "strange" to me:

```rust
// This line of code is just about width 100
fn evaluate_something_test(something: &Something, number: usize) -> Result<BTreeMap, CustomError> {
    // This comment is wrapped to 80. Lorem ipsum dolor sit amet, consectetur 
    // adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore 
    // magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation 
    // ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute...
    unimplemented!();
    
    // This looks strange to me. irure dolor in reprehenderit in voluptate 
    // velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint 
    // occaecat cupidatat non proident, sunt in culpa qui officia deserunt 
    // mollit anim id est laborum.
}
```

</details>

**Comment by calebcartwright** [comment]
will add explicitly, and more verbosely, that I'd prefer to see the default for this option be the same value as max_width as well. AFACIT the default value of 80 for this was established way back in 2015 in a rather ad-hoc manner during the earliest days of rustfmt development, and it seems to have continued on being the default ever since without any explicit discussion and reasoning around why it is less than max_width (at least none that I could find, though happy to be proven wrong).

**Comment by tgross35** [comment]
Would it be possible to default to whatever `max_width` is, instead of having a fixed default value?

I'm thinking that in most cases, anybody who changes `max_width` would probably want `comment_width` to change with it. Then explicitly setting `comment_width` would override that if needed

**Comment by c-git** [comment]
I remember looking up what the problem with stabilizing this was. But I can't remember where to find it. Can someone point me in the right direction please.

Since I'm commenting I'll just add a +1 for the single setting to `max_width` possibly with override as mentioned by tgross35 (Unless there is some reason why this is more complicated than it would seem).

**Comment by calebcartwright** [comment]
> Would it be possible to default to whatever max_width is, instead of having a fixed default value?

Yes, there's a number of width related options implemented this way

> I remember looking up what the problem with stabilizing this was. But I can't remember where to find it. Can someone point me in the right direction please.

More generally, see #5365 #5367

In this particular case, there's a salient, unresolved question of what the default value should be.

It's a question on the Style Team's radar as well, but full disclosure, I do not anticipate this is an option that'll be stabilized in the immediate future (i.e. I'd be surprised if this is done by year end, but suspect it will be by the time the 2024 edition drops)

**Comment by c-git** [comment]
Thank you appreciate the links. I wasn't sure where to look for them.

**Comment by calebcartwright** [comment]
> What's the current default value when I run cargo fmt on stable?

80
https://rust-lang.github.io/rustfmt/?version=v1.6.0&search=#comment_width

>  I would argue it shouldn't hold up stabilizing the mechanism for overriding that default.

I understand where you're coming from with this argument, but the default value for a config option (regardless of whether the option is a boolean, numeric, etc. type) is a critical factor and part of the long established process for stabilizing options.

Whether there could or should be changes to that process is fair topic for discussion, but so long as that is the process then that's what will be followed for all options, including this one.

**Comment by joshka** [comment]
> However, rustfmt doesn't format comments on stable, so this option isn't operative on stable

Is the following stronger restatement of this correct?

Given that this is an unstable option, the expectation of people that choose to use this option should be that formatting can and will change between versions. This expectation here implies that the default of 80 should not be relied upon by anyone to be a meaningful "default". To have a stable expected default on an unstable option is a contradiction on its face.

---

One thing that isn't a part of this option unlike the other options that is in many of the other line width options is that this option is unaffected by the [`use_small_heuristics`](https://rust-lang.github.io/rustfmt/?version=v1.8.0&search=#use_small_heuristics) option. Like others above, I'm a big fan of 100 char comments, as 80 characters seems too constrained to generally read well written prose, and I think strongly this should be the default.

I'd like to propose that the default should be based on 100% of max_width when `use_small_heuristics` option is set to either `Default` or `Max` (and 100, when it's set to `Off` and `max_width` is some other number).

**Comment by teohhanhui** [comment]
Realistically speaking, the default of 80 can't be changed easily without causing huge formatting changes across the entire ecosystem. So it's much more likely that it's already "too late" to change that default.

**Comment by waynr** [comment]
> It's a question on the Style Team's radar as well, but full disclosure, I do not anticipate this is an option that'll be stabilized in the immediate future (i.e. I'd be surprised if this is done by year end, but suspect it will be by the time the 2024 edition drops)

Does this comment mean that this config option won't be stabilized until the Style Team does something about it?

I'm curious because I occasionally run up against this and a couple other unstable features that I would like to have in rustfmt, but I don't want to be forced onto the unstable toolchain as a whole.

As for the default value of this option, I don't care what it is -- I just want to be able to set it myself. I would set it to 80 for my personal projects just like I do `max_width`. The default could be 20k for all I care, especially if the `wrap_comments` config is set to `false` by default.

I happen to be unemployed for now so if all this issue needs is someone to take the time to do something about it, I would be happy to take it on. It seems to me like most people agree that it should default to whatever `max_width` is, which would be great.

**Comment by joshka** [comment]
@waynr there's some recent info on team review bandwidth in https://rust-lang.zulipchat.com/#narrow/channel/357797-t-rustfmt/topic/PR.20for.20issue.206631.20has.20been.20open.20for.20two.20months/with/612803100 that probably applies to this just as much as any other part of rustfmt

**Comment by c-git** [comment]
@joshka thank you for linking the zulip thread

**Comment by waynr** [comment]
Thanks @joshka! It sounds like stabilizing unstable features probably isn't on the foreseeable horizon, maybe to be revisited after the team & community have draw down the overabundance of maintenance issues/PRs.
